### PR TITLE
Globalize: Add (Unicode Bidi Algorithm) compliant Bidi engine.

### PR DESCRIPTION
**Pull request by ashensis**
As per https://github.com/jquery/globalize/issues/542 discussion I submit the Bidi engine based on last Unicode Bidi character classification generated from "https://github.com/mathiasbynens/unicode-8.0.0"

### CODE REVIEW

**Comment by ashensis** [comment]
I do not understand what is wrong with CLA signature, I signed up multiple times with name: ashensis and emal: sashash@il.ibm.com and still this CLA check fails. My submission (by ashensis) should match the CLA signature.

**Comment by jzaefferer** [comment]
[The check results page](http://contribute.jquery.org/CLA/status/?repo=globalize&sha=07a040f757b29fd35cf373b9d152e3ecadd0af24) says that the name on the signature doesn't match the name on the commit. Looks to me like your signature is fine, but you're using a short/nickname on the commit. You can find instructions for updating the commit on the same page, though we can also take care of updating the commit author when merging your PR, since there is a valid signature already.

**Comment by ashensis** [comment]
Indeed I use the short name "ashensis" for my submissions, which constitutes my ID with jQuery. I have feeling that once I got signed CLA with my full name and the same email, my further attempts to sign with short name and the same email fail. I would prefer to use "ashensis" name for my further contributions. Could you please help me with resetting the CLA to "ashensis" with the same email sahsash@il.ibm.com. Thanks in advance.

**Comment by ashensis** [comment]
Thank you jzaefferer, I followed your advice and updated the submission author

**Comment by tomerm** [comment]
@rxaviers did you have a chance to review this submission ? Thx.

**Comment by rxaviers** [comment]
Hi, not yet sorry for the delay. Will do no longer than this coming week.

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,12 @@
...
+vendor
+NOTES
+.DS_Store
+.bundle
+Gemfile.lock
+.rvmrc
+*.sw*
+*.gem
+.rbx
+.ruby-version
```
Comment (rxaviers) [inline .gitignore]: These files are not included by Globalize, therefore they shouldn't be listed here.

File: .travis.yml [hunk 2]
```diff
@@ -1,9 +1,25 @@
...
+  # - jruby-19mode
+matrix:
+  allow_failures:
+    - rvm: rbx-2
+    - env: RAILS_4_2=true DB=sqlite3
+    - env: RAILS_4_2=true DB=postgres
+    - env: RAILS_4_2=true DB=mysql
+notifications:
+  webhooks:
+    - https://webhooks.gitter.im/e/40dc875af26f64f39090
```
Comment (rxaviers) [inline .travis.yml]: ?

**Comment by rxaviers** [comment]
@ashensis @tomerm please could you double check if this PR is what you intended to do? It's seems a completely different project has been submitted. It's ruby.

**Comment by tomerm** [comment]
@rxaviers I am not sure about all files. I am sure Alex can explain better. But at least those 3 are definitely relevant to Bidi engine: 

src/bidi-engine.js
src/bidi/unicode-types.js
src/build/intro-bidi-engine.js

**Comment by ashensis** [comment]
Indeed, as @Tomer said , these 3 files + Gruntfile.js corresponding changes are relevant. As for discrepancies in .gitignore .travis.yml AUTHORS.txt CONTRIBUTING.md package.json README.md and few others (fromtest\" I really don't remember how did the crall in, they most certainly have been changed with my last git pull synchronization, my apologies. Now got them updated to correct versions.

File: src/bidi-engine.js [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+			case "LRO":
+			case "RLO":
+			case "PDF":
+				_lastArabic = false;
+				break;
+			case "BN":
+				cType = "N";
+				break;
+		}
+		return cType;
```
Comment (rxaviers) [inline src/bidi-engine.js]: Are the above rules available in CLDR somewhere? I'm wondering if they are perhaps available in XML only.

Comment (rxaviers) [inline src/bidi-engine.js]: For clarity, the above sounds to me as the "[plural](https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/plurals.json)" rules for bidi.

Comment (rxaviers) [inline src/bidi-engine.js]: Nevermind, the above rules are for cTypes not locale parts.

**Comment by rxaviers** [comment]
Codewise, could you please include unit and functional tests? For now, a couple of unit tests are sufficient to help me understand the use cases / how this feature is used. Other tests could be included later along with the review. Feel free to push new commits including it incrementally.

**Comment by rxaviers** [comment]
Documentation is another piece I'm looking for: please update README.md including the module (and feel free to create `doc/*` files as needed). Just like the item above, feel free to push new commits incrementally implementing this.

**Comment by rxaviers** [comment]
Is this implementation (or part of it) based on any public specification like http://unicode.org/reports/tr9/?

**Comment by rxaviers** [comment]
There are coding styles changes to be made, but I don't want to nitpick for now. The above points are the major ones I see for now.

Thanks for your work so far.

File: src/bidi-engine.js (hunk 3 above)
Comment (ashensis) [inline src/bidi-engine.js]: Above mentioned code snippet represents an algorithmic part, not be expected in CLDR

Comment (rxaviers) [inline src/bidi-engine.js]: Yeap, I realized that. Nevermind.

**Comment by tomerm** [comment]
@rxaviers Indeed, this implementation is compliant with UBA (http://unicode.org/reports/tr9/).

**Comment by rxaviers** [comment]
> @rxaviers Indeed, this implementation is compliant with UBA (http://unicode.org/reports/tr9/). 

Excellent, so let's include source comments referencing parts of the specification where possible.

**Comment by ashensis** [comment]
Rafael, I will try to attend all requests above in near feature, as for contributing to documentation folder we have to make sure that this doesn't contradict IBM license. We have some concerns and are planning to clarify this.

**Comment by ashensis** [comment]
Hi Rafael, I got committed the following updates as requested:
1. Add documentation doc\api\bidi-engine\bidi-engine.md and updated Readme.md with relevant info
2. Add unites test\unit\bidi-engine\reordering.js
BTW, I don't realize the difference between unitest and functional test, so I pushed whole extensive testing under unites.
3. Update source file with appropriate references on relevant parts of http://unicode.org/reports/tr9/
Could you please have a look.

File: README.md [hunk 4]
```diff
@@ -364,6 +364,24 @@ handle dependencies and CLDR loading manually yourself.
...
+- **`bidiEng.doBidiReorder( [input]sourceText, [output, optional]sourceToTargetMap, [output, optional]levels ) `**
+
+  Returns reordered text.
+  
+  ```javascript
+  bidiEngine.doBidiReorder( "a(b)cA<B>C&123" ); 
+  // > "123&C<B>Aa(b)c" (upper case stands for Arabic or Hebrew)
+
+  [Read more...](doc/api/bidi-engine/bidi-engine.md)
+
```
Comment (rxaviers) [inline README.md]: Something is wrong with the markup. Click 'View' above and you'll note. I think it's a minor \`.

File: doc/api/bidi-engine/bidi-engine.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+var bidiEngine = Globalize.bidiEngine( );
+var options = [];
+options.isInputVisual = false;
+options.isOutputVisual = true;
+options.isInputRtl = true;
+
+bidiEngine.setOptions( options );
+
+var ret = bidiEngine.doBidiReorder( "Input text" );
```
Comment (rxaviers) [inline doc/api/bidi-engine/bidi-engine.md]: Something is wrong with this markup too. Click 'View' above and you'll note. Perhaps, another minor thing.

File: Gruntfile.js [hunk 6]
```diff
@@ -509,6 +509,25 @@ module.exports = function( grunt ) {
...
+								"./plural"
+							],
+							create: true,
+							override: {
+								wrap: {
+									startFile: "src/build/intro-bidi-engine.js",
+									endFile: "src/build/outro.js"
+								}
+							}
 						}
```
Comment (rxaviers) [inline Gruntfile.js]: There's a task called `uglify`, please include bidi-engine there. You can use one of the modules, e.g., date as baseline.

**Comment by rxaviers** [comment]
Thanks for the updates. The overall is clearer to me now. Will go through the details with more time. Sorry for the delay.

Friendly reminder for @tomerm: any updates for https://github.com/jquery/globalize/issues/542#issuecomment-171463089, thanks.

File: src/bidi-engine.js [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+		if ( levels ) {
+			_computeLevels( charArray, levels, { hiLevel: _dir } );
+		}
+		charArray.reverse();
+		sourceToTargetMap && sourceToTargetMap.reverse();
+		return charArray.join( "" );
+	};
+/**
+ * for reference see 3.3 in http://unicode.org/reports/tr9/
+*/
```
Comment (rxaviers) [inline src/bidi-engine.js]: Some coding styles nitpick... We only use `/**/` style for headers, even long comments use `//` http://contribute.jquery.org/style-guide/js/#comments. Please, could you update these comments using `//` instead. Also indent them according to the below code.

File: src/bidi-engine.js [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+	var _computeLevels = function( chars, levels, params ) {
+		var len = chars.length, prevState, newState = 0, newLevel, action, condition,
+			stateTable = _dir ? _STATE_TABLE_RTL : _STATE_TABLE_LTR,
+			condPos = -1, i, idx, types = [], resolvedTypes = [];
```
Comment (rxaviers) [inline src/bidi-engine.js]: Please, can you keep each variable in its own line sorted alphabetically? You could group all the ones that aren't initialized (e.g., `var action, condition, i, idx, ..., prevState, ...` ok to be in the same line)

Comment (rxaviers) [inline src/bidi-engine.js]: Is `idx` `index`? Can we use the full `index` name?

File: src/bidi-engine.js [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+		_lastArabic = false;
+		_hasUbatAl = false;
+		_hasUbatB = false;
+		_hasUbatS = false;
+		for ( i = 0; i < len; i++ ) {
+			types[i] = _getCharType( chars[i] );
+		}
+		for ( idx = 0; idx < len; idx++ ) {
+			prevState = newState;
+			resolvedTypes[idx] = _resolveCharType( chars, types, resolvedTypes, idx );
```
Comment (rxaviers) [inline src/bidi-engine.js]: Note spaces `resolvedTypes[idx]` -> `resolvedTypes[ idx ]`

File: src/bidi-engine.js [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+ * undefined(Contextual direction, i.e.direction defined by first strong character of input string)
+ * - isOutputVisual [boolean] (defaults to false):
+ * allowed values: true(Visual mode), false(Logical mode)
+ * - isOutputRtl [boolean]:
+ * allowed values true(Right-to-left direction), false (Left-to-right directiion),
+ * undefined(Contextual direction, i.e.direction defined by first strong character of input string)
+ * - isSymmetricSwapping [boolean] (defaults to false):
+ * allowed values true(needs symmetric swapping), false (no need in symmetric swapping),
+ */
+	this.$bidiEngine.setOptions = function( options ) {
```
Comment (rxaviers) [inline src/bidi-engine.js]: Why `bidiEngine` has a `$` in the name please?

File: src/bidi-engine.js [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+			_isOutRtl = options.isOutputRtl;
+			_isSymmetricSwapping = options.isSymmetricSwapping;
+		}
+	};
+
+	this.$bidiEngine.setOptions( options );
+	return this.$bidiEngine;
+};
+
+Globalize._bidiUnicodeTypes = bidiUnicodeTypes;
```
Comment (rxaviers) [inline src/bidi-engine.js]: Is `Globalize._bidiUnicodeTypes` accessed outside of this file?

File: test/unit/bidi-engine/reordering.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  options.isInputRtl = false;
+  options.isOutputRtl = true;
+  options.isSymmetricSwapping = true;
+
+  bidiEngine.setOptions( options );
+
+  var reorderedString = bidiEngine.doBidiReorder( sourceString , sourceToTargetMap, levels );
+  assert.equal( reorderedString, "\u05d0>\u05d5<\u05ea& 321c)b(a" );
+  assert.deepEqual( sourceToTargetMap, [5,6,7,8,9,10,11,14,13,12,4,3,2,1,0] );
+  assert.deepEqual( levels, [0,0,0,0,0,1,1,1,1,1,1,1,2,2,2] );
```
Comment (rxaviers) [inline test/unit/bidi-engine/reordering.js]: Spaces `[0, 0, 0, ...`

File: src/bidi-engine.js (hunk 11 above)
Comment (ashensis) [inline src/bidi-engine.js]: Sorry, will remove from Globalize scope

File: src/bidi-engine.js (hunk 10 above)
Comment (ashensis) [inline src/bidi-engine.js]: I wanted to signify the exposed scope by this, if you don't like this, I can remove '$'

File: README.md (hunk 4 above)
Comment (ashensis) [inline README.md]: Can't wrap my head around, could you clarify please, what is wrong?

Comment (Arkni) [inline README.md]: @ashensis here are some notes:

1. Headers

  this:
  ```md
  - ** Globalize.bidiEngine( opt ) `**
  ```
  should be:
  ```md
  - **`Globalize.bidiEngine( opt )`**
  ```
2. JS snippets:

  this
  ```
  ```javascript
  var bidiEngine = Globalize.bidiEngine( options );
  ```
  should be:
  <pre>
  ```javascript
  var bidiEngine = Globalize.bidiEngine( options );
  ```
  </pre>
  The same thing for the other snippet.

File: doc/api/bidi-engine/bidi-engine.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+```javascript
+  var options = {
+	isInputVisual: true,
+	isInputRtl: false,
+	isOutputVisual: false,
+	isOutputRtl: true,
+	isSymmetricSwapping: true
+  };
+  var bidiEngine = Globalize.bidiEngine( options );
+
```
Comment (Arkni) [inline doc/api/bidi-engine/bidi-engine.md]: You should always close the snippets. Add \`\`\` in a new line after `var bindingEngine = ...`

File: doc/api/bidi-engine/bidi-engine.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+Output optional parameter - Array containing Bidi levels of characters in source string
+
+  ```javascript
+  var sourceToTargetMap = [], levels = [];
+  var bidiEngine = Globalize.bidiEngine( {isInputVisual: true, isSymmetricSwapping: true} );
+  var reorderedString = bidiEngine.doBidiReorder( "a(b)cA<B>C& 123", sourceToTargetMap, levels ); 
+  // reorderedString > "123 &C<B>Aa(b)c" (upper case stands for Arabic or Hebrew)
+  // sourceToTargetMap >
+  // levels >
+
```
Comment (Arkni) [inline doc/api/bidi-engine/bidi-engine.md]: The same as the above notice: you need to add \`\`\` to the next line after `// levels >`

File: doc/api/bidi-engine/bidi-engine.md (hunk 5 above)
Comment (Arkni) [inline doc/api/bidi-engine/bidi-engine.md]: The same as the above comments. Add \`\`\` in a new line after `var ret = b ...`

File: src/bidi-engine.js [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+
+	var _lastArabic = false, _hasUbatAl, _hasUbatB, _hasUbatS, DIR_LTR = 0, DIR_RTL = 1,
+		_isInVisual, _isInRtl, _isOutVisual, _isOutRtl, _isSymmetricSwapping, _dir = DIR_LTR;
+
+	this.$bidiEngine = {};
+
+	var _init = function( text, sourceToTargetMap ) {
+		if ( sourceToTargetMap ) {
+			for ( var i = 0; i < text.length; i++ ) {
+				sourceToTargetMap[i] = i;
```
Comment (Arkni) [inline src/bidi-engine.js]: spaces
`sourceToTargetMap[i]` => `sourceToTargetMap[ i ]`

File: README.md (hunk 4 above)
Comment (ashensis) [inline README.md]: Thanks a lot for clarification, @Arkni

**Comment by ashensis** [comment]
In last commit I tried to attend all remarks as follows:
1. added bidi engine to uglify task to produce minimal version
2. replaced idx by index everywhere, changed variables declaration, 
added spaces before and after content of square brackets.
3. removed _bidiUnicodeTypes from Globalize scope, removed '$' from exposed namespace.
4. added spaces in resultant arrays in reordering.js unitest module.
5. changed /* comments with //
6. changed markup in document files

**Comment by tomerm** [comment]
@rxaviers Any outstanding comments on this PR ?

**Comment by rxaviers** [comment]
Hi, @jzaefferer and I have both reviewed the PR.

The diagnostic we have is similar to the one for the Arabic Shaping feature. The implementation helped us to understand the dependencies the module has on Globalize and it turns out there is almost none. It means the module is loosely coupled with Globalize. With that in mind, we suggest you create a standalone BiDi engine library. You should be able to take the code from this PR, which is already complete, and provide the bits of infrastructure around that you actually need (which is likely less then what Globalize needs), so I believe the investment you made so far by implementing this feature can be fully utilized in a standalone library. You can then update that module independently of Globalize. And it will be easier to maintain, since you won’t have to support globalize-compiler (which doesn’t provide any benefit for this engine).

Having said that, if you want to make this a jQuery hosted project as well, I believe @kborchers will be happy in assisting you further too.

**Comment by rxaviers** [comment]
Please, just let us know if you have additional questions.

**Comment by tomerm** [comment]
@rxaviers thanks a lot for assistance and advise. 

@kborchers I believe it does make sense to have Bidi engine library (JS implementation of Unicode Bidi Algorithm - http://unicode.org/reports/tr9/) to be a JQuery hosted projects. I would appreciate your guidance for next steps. Many thanks in advance.

**Comment by kborchers** [comment]
@tomerm - as we work to finalize a new technical organization and support structure for the Foundation, we are defining policies, processes and procedures for current and new projects at https://github.com/jquery-foundation/TAC. I would recommend reading through that repo and then do a couple of things:

1. Let us know if anything needs clarification, more detail, etc by filing issues
2. Follow the new project application process outlined in the Project-Lifecycle.md and again, if things are unclear or need further explanation, please file issues to help us improve those documents

Also, please keep in mind that we are still finalizing that structure so some parts are not completely in place like the members of the TAC for example. We hope to bring all of that together soon but until then, there will likely be delays in working projects into the Foundation but we hope to continually improve and streamline that process as we go. This should not discourage you from starting this new project under your personal GitHub org to keep everything moving and then work toward joining the Foundation. Thanks!

**Pull request closed without merging**